New tutorial

andrewwinters5000 opened the pull request:
This adds an additional tutorial and discussion of the newly released feature in HOHQMesh that one can define a straight about which a mesh can be reflected.

andrewwinters5000 commented:
<img width="993" alt="Screenshot 2024-06-14 at 23 28 43" src="https://github.com/trixi-framework/HOHQMesh.jl/assets/25242486/6712cbe8-d7bf-4142-99af-dc032f26e768">


<img width="991" alt="Screenshot 2024-06-14 at 23 28 59" src="https://github.com/trixi-framework/HOHQMesh.jl/assets/25242486/daa915fa-11ac-411e-950c-e31958494df3">

<img width="992" alt="Screenshot 2024-06-15 at 00 06 29" src="https://github.com/trixi-framework/HOHQMesh.jl/assets/25242486/2672e82d-6512-491a-9ff6-4ce3fd06da88">

<img width="993" alt="Screenshot 2024-06-15 at 00 06 43" src="https://github.com/trixi-framework/HOHQMesh.jl/assets/25242486/9e416528-dc42-465b-b874-3641bb0d1e7e">

DavidAKopriva commented:
A couple of comments:
1. HOHQMesh (rather than fortran) is case insensitive for the "symmetry" keyword. (Hmm, but all other keywords *are* case sensitive.
2. You should only need to change the curve names rather than re-construct the model. e.g. setCurveName!(line9,"B9") and setCurveName!(line2,"symmetry") and then regenerate.
3. In the second tutorial, the boundaries O.4 and O.8 will be B3 and B3_R and similarly for O.8. They become internal boundaries, rather than internal parts of the mesh. To do a fully symmetric mesh without internal boundaries, O.4 and O.8 need their names to *also* be set to "symmetry". This flexibility lets one define internal boundaries, as requested in issue 49 of HOHQMesh.
4. You need to delete the mesh before re-generating. (I think.) I suppose the default behavior should be that when generate_mesh is called the mesh, if it exists, should be deleted. Does it do that? I don't remember.

That's all I can think of at the moment.

sloede commented on docs/src/tutorials/symmetric_mesh.md:
```diff
@@ -0,0 +1,416 @@
+as the project. For example, the resulting HOHQMesh control file from this tutorial
+will be named `symmetric_mesh.control`.
+If the folder `out` does not exist, it will be created automatically in
+the current file path.
```
```suggestion
the current working directory.
```
Clearer to me, but YMMV 🤷‍♂️

andrewwinters5000 commented:
Thanks for the comments @DavidAKopriva ! I will update and simplify the second part of the tutorial from your suggestions. I also need to investigate the case sensitivity (or seemingly the lack of it) for the boundary name `"symmetry"`. I tried a the three capitalization strategies for `"Symmetry"`, `"SYMMETRY"` and `sYmMeTrY"` and a valid mesh was generated for all of them.

DavidAKopriva commented:
Thanks for writing the tutorial. Yes, HOHQMesh calls ToLower(str) on the name before testing if it is equal to "symmetry", so it doesn't matter what case one uses. We should think if that should be done for all keywords, but it doesn't matter using HOHQmesh.jl because the control file written out will have the correct case.

andrewwinters5000 commented:
> Thanks for writing the tutorial. Yes, HOHQMesh calls ToLower(str) on the name before testing if it is equal to "symmetry", so it doesn't matter what case one uses. We should think if that should be done for all keywords, but it doesn't matter using HOHQmesh.jl because the control file written out will have the correct case.

Thanks for the information. I assumed that there was something under hood helping with the case sensitivity, but I was only checking inside the Julia code. Avoiding case sensitivity on all the keywords could help make the control files more " idiot proof", but we can discuss this further in a different HOHQMesh PR.

andrewwinters5000 commented:
New figures

![Screenshot 2024-06-19 at 12 45 05](https://github.com/trixi-framework/HOHQMesh.jl/assets/25242486/542fa4dd-14f8-4643-b2cc-2fbe09d1da51)
![Screenshot 2024-06-19 at 12 45 35](https://github.com/trixi-framework/HOHQMesh.jl/assets/25242486/4cad99b8-2437-4687-8dbd-2835e100cb8f)
![Screenshot 2024-06-19 at 12 45 58](https://github.com/trixi-framework/HOHQMesh.jl/assets/25242486/fb806e49-0f9a-402e-8613-3c5ae5452b9e)
![Screenshot 2024-06-19 at 12 46 17](https://github.com/trixi-framework/HOHQMesh.jl/assets/25242486/95a0c364-9120-406e-9a94-d47532f01752)

andrewwinters5000 commented:
Better boundary name figures

![Screenshot 2024-06-19 at 14 21 30](https://github.com/trixi-framework/HOHQMesh.jl/assets/25242486/d332e86a-f958-4362-81b9-2ad40a408d94)
![Screenshot 2024-06-19 at 14 21 53](https://github.com/trixi-framework/HOHQMesh.jl/assets/25242486/1afd159e-e8c4-459f-ba7d-a31057b0c135)
![Screenshot 2024-06-19 at 14 22 17](https://github.com/trixi-framework/HOHQMesh.jl/assets/25242486/314a7bc7-0cb1-4708-bbb9-98ca85798d5b)
![Screenshot 2024-06-19 at 14 22 44](https://github.com/trixi-framework/HOHQMesh.jl/assets/25242486/2476b6fe-79a4-4d3a-98b1-089714c76ade)

DavidAKopriva reviewed (changes requested):
Changes as is are fine, but I think the renameCurve! procedure description should be added to the API documentation, probably in the model API since it is a model operation. I don't think that it is in there yet.

andrewwinters5000 commented:
> Changes as is are fine, but I think the renameCurve! procedure description should be added to the API documentation, probably in the model API since it is a model operation. I don't think that it is in there yet.

I added the `renameCurve!` to the API and updated the `remover_mesh!` function to also delete the plot, control, and stats files.

sloede commented on docs/src/tutorials/symmetric_mesh.md:
```diff
@@ -0,0 +1,360 @@
+in HOHQMesh that says it is the straight line across which
+a reflection will occur.
+
+!!! tip "Name of the symmetry boundary"
+    As noted above, `"symmetry"` is a keyword for the HOHQMesh generator that
+    prescribes over which line a reflection of the mesh occurs. The Fortran implementation
+    for this keyword in HOHQMesh is not case sensitive. So, within the
+    Julia script one can name this reflection boundary line `"symmetry"` or `"Symmetry"`
+    or `"SYMMETRY"` or even something strange like `"sYmMeTrY"` and the mesh will still
+    generate successfully.
```
Does it have to be such a widely and, imho, accidentally used name? Wouldn't it be possible to pick a name that would be invalid for a boundary name (e.g., "{symmetry}" or ":symmetry" etc.) such that no accidental use can happen? Or, if you want to keep it a valid boundary name for implementation reasons, make it something "unusual" such as "__symmetry__".

If "symmetry" it shall remain, I suggest to prominently document this in the main docs/tutorials to highlight that this is a special name.

andrewwinters5000 replied:
David and I discussed this name a bit and settled on "symmetry". However, making a specialized variant of the name to avoid accidental misuse is a good idea.

sloede commented on docs/src/tutorials/symmetric_mesh.md:
```diff
@@ -0,0 +1,360 @@
+in HOHQMesh that says it is the straight line across which
+a reflection will occur.
+
+!!! tip "Name of the symmetry boundary"
+    As noted above, `"symmetry"` is a keyword for the HOHQMesh generator that
+    prescribes over which line a reflection of the mesh occurs. The Fortran implementation
+    for this keyword in HOHQMesh is not case sensitive. So, within the
+    Julia script one can name this reflection boundary line `"symmetry"` or `"Symmetry"`
+    or `"SYMMETRY"` or even something strange like `"sYmMeTrY"` and the mesh will still
+    generate successfully.
```
As a second note, I'd recommend to not confuse users with implementation details. I'd just enforce the lower-case version in Julia and that's it 🤷‍♂️

If this is too much of a violation of your sense for consistency (which I would totally understand, with me being me 😬), just shorten it to something like
```suggestion
!!! tip "Name of the symmetry boundary is case-insensitive"
    The name of the reflection boundary line is not case-sensitive, thus`"symmetry"` or `"Symmetry"` or `"SYMMETRY"` are all valid and will be recognized as the keyword for symmetry.
```

andrewwinters5000 replied:
Sure, we can avoid getting users bogged down in such details.

sloede commented on docs/src/tutorials/symmetric_mesh.md:
```diff
@@ -0,0 +1,360 @@
+control file `symmetric_mesh.control`, plot file `symmetric_mesh.tec`,
+and mesh statistics file `symmetric_mesh.txt` from the `out` folder.
+
+To illustrate the reflection about multiple boundary curves (which must be co-linear!),
```
Is there a check for co-linearity? If not, it might be good to note that this is something that is not enforced and that the user is responsible for correct use.

If there is a check (which would make sense imho), it would be nice to inform the user about it as well here.

DavidAKopriva replied:
Yes, an exception is thrown and the user is informed that the reflection was not performed. The user is also informed if the curve along which the reflection is requested is not a straight line. The tests for the latter are not exhaustive. It basically checks to see if the second derivative (that code already existed) is zero at a couple of points.

sloede reviewed (commented):
Just a few comments/suggestions that came up when perusing this PR

andrewwinters5000 commented:
New figures now that that symmetry name has changed to `:symmetry`

<img width="803" alt="Screenshot 2024-06-21 at 23 43 08" src="https://github.com/trixi-framework/HOHQMesh.jl/assets/25242486/2755fb08-9796-4e79-9d2d-055ef476dd4b">
<img width="1045" alt="Screenshot 2024-06-21 at 23 43 26" src="https://github.com/trixi-framework/HOHQMesh.jl/assets/25242486/c271d208-6779-4e47-9045-a8cd21c5896d">
<img width="801" alt="Screenshot 2024-06-21 at 23 43 47" src="https://github.com/trixi-framework/HOHQMesh.jl/assets/25242486/0118a02f-0343-4280-8a63-eba5a59d69e3">
<img width="1063" alt="Screenshot 2024-06-21 at 23 44 05" src="https://github.com/trixi-framework/HOHQMesh.jl/assets/25242486/7e5baf73-9c48-45be-84e4-90eb331103a5">

DavidAKopriva reviewed (approved):
I think the changes are fine, and I approve modulo the windows fix.

sloede reviewed (dismissed):
LGTM

DavidAKopriva reviewed (approved):
I see there is no check on whether or not a file exists when rm() is called in the remove_mesh! procedure. I would guess that an error message will be written, and after thinking about it, that's OK.

andrewwinters5000 commented:
@sloede Are we good to merge? Once this is done I can release a new version of HOHQMesh.jl
